joomla/jissues
# Language flags

Ref #764 

https://github.com/joomla/jissues/issues/764#issuecomment-211984084
> could we use SVG's for the flags instead of separate GIF's that cause 18 additional resource loads per page

Now this opened not one but several cans of worms for me...

The most easy solution (I found) would be going to https://www.flag-sprites.com generate the combined flag image plus the required CSS and go home.

Unfortunately the flags they offer do not contain some languages we currently use. Mainly catalan and british english. We may leave those blank or just choose a "nearby" one but - using an EU image flag for en-GB might get us in deepest trouble (who knows if they really go with the brexit), and I don't want to get killed for using the "wrong flag"...

So I bit the bullet (before it hits me)... and my first stop was here: http://www.phpied.com/command-line-css-spriting which gave me some ideas on how this could be achieved.

### Language codes

Another thing to consider was the language "code" to use.
Currently we are using "Joomla! language codes&trade;". The only reference I found was at [Micro$oft](https://msdn.microsoft.com/en-us/library/ee825488(v=cs.20).aspx).

There are several other [language codes](https://en.wikipedia.org/wiki/Language_code) including

* [ISO language codes](http://www.lingoes.net/en/translator/langcode.htm)
* [ISO 639-X](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)
* [ISO 3166-1 alpha-2](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)
* [Crowdin language codes](https://crowdin.com/page/api/language-codes) (which seems to be the only one supporting [Klingon](https://en.wikipedia.org/wiki/Klingon_language) (and probably others) - now that's what I call future proof...)

There are also some PHP Libs dealing with language codes:

* https://github.com/matriphe/php-iso-639
* https://github.com/alcohol/iso3166

I decided not to use any of those but instead reinvent the wheel (again) - I love Joomla! you know - and added a list of known languages with "our" codes to the language helper class. See: https://github.com/joomla/jissues/pull/801/files#diff-fd55cb6560b16d21f62ec0d0898c7e17R43

### Flag images

This is the best and most complete flag images collection that I found and that is free of charge (as in 0$) available on the internet:

* http://forum.tsgk.com/viewtopic.php?t=4921

You really wont miss the [Jolly Rogers](https://en.wikipedia.org/wiki/Jolly_Roger) collection including a flag icon for [Calico Jack Rackham](https://en.wikipedia.org/wiki/Calico_Jack): ![calico jack rackham](https://cloud.githubusercontent.com/assets/33978/16248392/82759812-37d4-11e6-979d-6b74e57423d7.png)

### Test instructions

* There is a list of known languages which is copied from here: https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes
    * Translators: Please review the code [here](https://github.com/joomla/jissues/pull/801/files#diff-fd55cb6560b16d21f62ec0d0898c7e17R43)  and check that the name that will be displayed in the language selector (the `display` property) is **correct**
* There is a new command `jtracker make languageflags`. This command requires [imagemagick](https://www.imagemagick.org) to be installed on your system and the flag images to be stored "somewhere". To test this command you have to download the flag images from [here](http://forum.tsgk.com/viewtopic.php?t=4921) as they are not included in this PR.<br />
This command will do the following:
    1. Create a combined image with flags for the languages currently "installed" in our `JTracker` application using the [montage]() command and store it at the path given by the `--imagefile` option (defaults to `/cache/flags.png`)
    1. Generate a CSS file at the path given by the `--cssfile` option (defaults to `/cache/flags.css`).
* Hopefully you have a local installation so you can check that the language selector displays and works correctly:

Before | After
--- | ---
![2016-06-21-161157_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/16248787/18b7b042-37d7-11e6-8154-75c8851d8473.png) | ![2016-06-21-160858_1366x768_scrot](https://cloud.githubusercontent.com/assets/33978/16248791/1cc03e84-37d7-11e6-964a-eaf591ef76b4.png)


#### hF
